lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp

```diff
@@ -2511,162 +2181,510 @@ class StableHLOToTTIRLogicalAndBitwiseOpConversionPattern
+      std::make_error_code(std::errc::operation_not_supported));
+}
+
+static llvm::ErrorOr<ttcore::ReduceType> getReduceType(Operation *op) {
+  if (!llvm::isa<mlir::stablehlo::AllReduceOp, mlir::stablehlo::ReduceScatterOp,
+                 mlir::stablehlo::ScatterOp>(op)) {
+    return llvm::ErrorOr<ttcore::ReduceType>(
+        std::make_error_code(std::errc::operation_not_supported));
+  }
+  return getReduceTypeFromRegion(op->getRegion(0));
+}
+
+static LogicalResult
+determineClusterAxis(::mlir::DenseIntElementsAttr replicaGroups,
+                     uint32_t &clusterAxis) {
+  /*
+  We need to figure out what the cluster axis is based on replica_groups.
+  Replica groups define which device axis we are performing the collective
+  communication operation on. It is a 2D vector. Each element in replica_groups
+  contains a list of devices that will perform the collective communication
+  operation with each other. Currently we only support 2D meshes, but this
+  algorithm can be expanded for ND.
+
+  ex.
+  mesh = [2, 4]
+  replica_groups = [[0, 1, 2, 3], [4, 5, 6, 7]]
+  0 1 2 3
+  4 5 6 7
+
+  collective communication operation happens on (0, 1, 2, 3) and (4, 5, 6, 7) so
+  cluster_axis = 1 (mesh[1])
+
+  mesh = [2, 4]
+  replica_groups = [[0, 4], [1, 5], [2, 6], [3, 7]]
+  0 1 2 3
+  4 5 6 7
+
+  collective communication operation happens on (0, 4), (1, 5), (2, 6), (3, 7)
+  so cluster_axis = 0 (mesh[0])
+
+  */
+  auto replicaGroupsShape = replicaGroups.getType().getShape();
+
+  if (replicaGroupsShape.size() != 2) {
+    // Can only have replica groups of size 2. Otherwise, this is an ill formed
+    // graph and needs to be asserted.
+    return failure();
+  }
+
+  // Case where we have single devices in each replica_group (ie perform
+  // collective communication operation against itself which should be optimized
+  // away). We also assume we are only using our constrained mesh types (ie 1x8,
+  // 1x32 etc) and cannot have (32x1, 8x1).
+  if (replicaGroupsShape[1] != 1) {
+    auto firstElementIt = replicaGroups.begin();
+    auto secondElementIt = firstElementIt + 1;
+
+    clusterAxis = (((*firstElementIt) + 1) == *secondElementIt);
+    return success();
+  }
+
+  // If replicaGroupsShape[1] == 1, then the cluster axis should be set to 0.
+  clusterAxis = 0;
+  return success();
+}
+
+// StalbeHLO spec.md defines following channel type for ccl ops
+enum StableHLOChannelType {
+  // CHANNEL_TYPE_INVALID = 0 : Invalid primitive type to serve as
+  // default.
+  kChannelTypeInvalid = 0,
+  // DEVICE_TO_DEVICE = 1 : A channel for sending data between
+  // devices.
+  kChannelTypeDeviceToDevice = 1,
+  // DEVICE_TO_HOST = 2 : A channel for sending data from the
+  // device to the host. Can only be used with a Send operation.
+  kChannelTypeDeviceToHost = 2,
+  // HOST_TO_DEVICE = 3 : A channel for sending data from the host to
+  // the device. Can only be used with a Recv operation.
+  kChannelTypeHostToDevice = 3,
+};
```

Comment:
:warning: **llvm\-else\-after\-return** :warning:
do not use `` else `` after `` return ``
```suggestion
    } if (isa<mlir::stablehlo::MulOp>(op)) {
```